Source/WebKit/UIProcess/WebPageProxy.cpp

```diff
@@ -9090,6 +9090,13 @@ void WebPageProxy::queryPermission(const ClientOrigin& clientOrigin, const Permi
         // this topOrigin has requested permission to use the Notifications API previously.
         if (m_notificationPermissionRequesters.contains(clientOrigin.topOrigin))
             shouldChangeDeniedToPrompt = false;
+
+        // Always return "denied" in ephemeral browsing sessions, for consistency with
+        // `Notification.permission` and `Notification.requestPermission`.
+        if (sessionID().isEphemeral()) {
```

Comment:
I think what we want here is something like:
if (sessionID().isEphemeral()) {
     completionHandler(shouldChangeDeniedToPrompt ? PermissionState::Prompt : PermissionState::Denied)
     return;
}
This mimics the fact that the UI delegate would always return denied for ephemeral session.
This would change the above test to expect "prompt".

We would then update the Notification.permission code path to match navigator.permissions.query.
This might mean partially reverting https://github.com/WebKit/WebKit/commit/0570dfb680f59ef121a7faa71d1e8e6297f3594f, in particular the changes in WebNotificationClient::checkPermission.

Reply:
Sounds good — I'll unify the behaviors (but in the opposite direction, so that `Notification.permission` also returns prompt in this case).